Add preamble and example to pattern matching in definitions

jakubsvec001 opened the pull request:
@cuihtlauac 

As I work my way though your tutorials, I am happy to make little updates and tweaks like the present one and the ones we've worked through in the last few days, but I want to make sure I'm not swamping you with extra work.

Over the last year I've been studying Category and Type Theory, with a notable focus on Grothendieck Toposes. I am currently refreshing my knowledge in OCaml, and want to improve my fluency and accuracy when discussing the programming language concretely and functional programming in the abstract. By participating in pull requests of the tutorials, I am able to practice this. 

But I want to check in with you and your team to ensure that I am not giving you additional overhead when checking these pull requests. I'd much rather align with your goals.

If you have any thoughts or suggestions, please let me know.

For example, if you would like me to only focus on typos I can restrict myself to those. A slightly wider scope would include disambiguation, and minor expansion or preambles on topics. For an even wider scope, if you don't mind I take a stab at modifying a document when I find a `FIXME` (like the present pull request),  I'd greatly appreciate it (as well as any feedback you have as we refine it). 

Finally, if you have any larger tasks for your tutorials that you've been ruminating on but not had the chance to work on, I'd be happy to work on them if you can itemize and prioritize them.

jakubsvec001 committed:
Add preamble to pattern matching section
Add nested pattern matching example, update discarding pattern matched value examples

bbatsov commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,103 @@ val forename : string = "Robin"
+Pattern matching also works with nested user-defined types.
+
+In the example below, we deconstruct nested tuples:
+```ocaml
````
Generally it's a good idea to leave a blank line between code blocks and the surrounding text.

bbatsov replied:
(even if the code in the guides doesn't do this consistently)

bbatsov commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,103 @@ val forename : string = "Robin"
+```
+In the following example, we deconstruct nested records and an associated tuple.
+
+First, define an address record type
````
Probably such sentences should end with `:`.

bbatsov commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,103 @@ val forename : string = "Robin"
+val email : string = "john@example.com"
+val phone : int = 1234567890
+```
+**note**: `address` and `contact` are not defined as variables available in your scope. Instead, they serve as binding patterns that help destructure the nested structure.
````
note -> Note, to be consistent with the rest of the docs.

jakubsvec001 commented:
Updated based on bbatrov's suggestions:

- added blank lines around code blocks
- fixed typos

Thank you, @bbatsov, for taking the time to review

cuihtlauac commented:
Thanks @jakubsvec001 and @bbatsov 

Can you move spacing into another PR? Single-purpose patches are more straightforward to review. 

> But I want to check in with you and your team to ensure that I am not giving you additional overhead when checking these pull requests. I'd much rather align with your goals.

We love contributions. Send you what you feel is needed to improve the materials. Fresh eyes are precious. It's an opportunity to catch overlooked issues. 

> Finally, if you have any larger tasks for your tutorials that you've been ruminating on but not had the chance to work on, I'd be happy to work on them if you can itemize and prioritize them.

Both fixing typos and providing full tutorials make great contributions. We still need tutorials on `Format`, polymorphic variants (PR #1531 could be resurrected), laziness, GADTs, and others. Small tutorials are also great. Refreshing old tutorials, too.

And if you can illustrate what a topos can be using OCaml and without math terminology, we'd love that too ;-)

jakubsvec001 commented:
@cuihtlauac, I `git reset --hard HEAD~1` and will send the removed changes in another pull request

jakubsvec001 commented:
> Both fixing typos and providing full tutorials make great contributions. We still need tutorials on Format, polymorphic variants (PR https://github.com/ocaml/ocaml.org/pull/1531 could be resurrected), laziness, GADTs, and others. Small tutorials are also great. Refreshing old tutorials, too.

Great. I'll work from the bottom up, beginning with typos and minor updates for streamlining/clarification, then look into new content after I've made it through the present material.

jakubsvec001 commented:
@bbatsov What is OCaml.org's style-guide preference for line length? Several .md files switch between no line-length restriction and ~80-85 character line breaks.

I've read the ocaml.org/DOCUMENTATION_WRITING.md file and don't see a reference to this.

In documentation, I have a preference for setting my IDE to wrap lines automatically based on window width over enforcing manual breaks. Updates to documentation using git without line-length breaks can be simpler.

HOWEVER, I'm happy to conform to your org's standards.

With your feedback, I can follow this in my updates. I can also make a specific pull request standardizing this throughout the `learn` documentation.

bbatsov commented:
> @bbatsov What is OCaml.org's style-guide preference for line length? Several .md files switch between no line-length restriction and ~80-85 character line breaks.

There are no markdown standards currently set for the repo, therefore there are many inconsistencies in the various docs. Probably at some point we'll need to setup `markdownlint`for the repo to start working in this direction. 

> With your feedback, I can follow this in my updates. I can also make a specific pull request standardizing this throughout the learn documentation.

That'd be nice I think. Personally I'm OK with line-length up to 80-100 characters, as most style guides choose one of the two. I agree it's generally a good idea to fold long lines, otherwise both edits and code reviews become harder. (not to mention it's also harder to read those as a human)

cuihtlauac commented:
It is accurate that there's no markdown style guide in ocaml.org. We're open to suggestions on this.

Using `markdownlint` sounds like a good idea. What are its dependencies?

bbatsov commented:
Depends on the implementation. The original was written in Ruby (see https://github.com/markdownlint/markdownlint) and there's a newer version using Node.js (https://github.com/DavidAnson/markdownlint). Both will get the job done, but the node.js is more popular these days and kind of a standard in the broader community.

jakubsvec001 commented:
@cuihtlauac @bbatsov I am new to contributing to a project as robust as this so am coming at this without preferences.

This morning, I played around with Markdownlint.js and Pandoc and pasted my notes below:

**Notes**:
- markdownlint.js has two different cli tools that need to be installed separately:
  1. markdownlint-cli is the most common tool for lint-checking files, but it does not automatically fix issues
  2. markdownlint-cli2 is a more advanced cli tool but it is not a replacement for markdownlint-cli. It provides a `--fix` flag for automaticaly applying fixes on fixable errors
- a .markdown.json configuration file can be used to customize behavior. The config file must be in the same directory the cli tool is called from or a -config flag and path/to/file must be provided. I wish the tool would recursively search parent directories for a config file but maybe that would be more frustrating for other scenarios. The issue from my perspective is that it gives a contributor one extra thing to think about so as not to run commands with incorrect configs.
- markdownlint.js does not support auto-fixing line breaks
- Pandoc has a line-length auto-fixing feature and seems robust (from my limited testing) in how it breaks/unbreak/rebreaks lines to a selected column-width

**WORKFLOW 1**: Contributor's responsibility
As an external contributor I expect to have to read a style guide and contribution criteria to conform to. I can also be expected to download any tools necessary for submission and run them. Suggestions for commons commands would be helpful in the style guide to set proper flags, either for finding the .markdownlint.json config or for supplying some configs through the command-line. 

In this way, adding this step to a contributor's workflow would solve your issue by formatting downstream of the pull request. A github pull workflow can then validate proper formatting.

**WORKFLOW 2**: Transparently apply fixes before incorporating pull request
However, the above adds friction to a contributor's workflow. Another option is to solve this problem transparently by having a github workflow perform automatic fixes upon a pull request being accepted using markdownlint-cli, cli2, and pandoc. This, however, there may be edge cases that cause some text being unexpectedly modified without being checked by a human.

**WORKFLOW 3**: A combination of 1 and 2
A final option is to do both.

I'm sure there are other considerations I'm not accounting for, but I wanted to look into it and wrap my head around some options.

If you would like, I'd be happy to write a first draft of a .markdown.json config file as well as a boilerplate style guide with instructions for downloading markdownlint.js and pandoc, along with commands to run.

bbatsov commented:
I like option 1, potentially combined with some CI job that checks modified markdown files and flags problems with them, so gradually over time we'd get consistent formatting across the board.

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -114,6 +114,11 @@ Arbitrary combinations of chaining or nesting are allowed.
 
+## Introduction to Pattern Matching
+
+[Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) is a mechanism in programming language design that enables the destructuring of values and facilitates case analysis in a concise and expressive manner.
```
```suggestion
[Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) is a programming language construct that generalizes case analysis, it makes subexpression inspection possible and applies to values of any type. 
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -114,6 +114,11 @@ Arbitrary combinations of chaining or nesting are allowed.
+
+[Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) is a mechanism in programming language design that enables the destructuring of values and facilitates case analysis in a concise and expressive manner.
+
+In the following sections, we explore definitional pattern matching using `let` bindings. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in the construction of conditional expressions using `match...with` and compare it to `if...then...else`. In the chapter on [Error Handling](/docs/error-handling), we explore how destructuring  ...
```
```suggestion
In the following sections, we explore matching inside `let` bindings. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in the construction of conditional expressions using `match...with` and compare it to `if...then...else`. In the chapter on [Error Handling](/docs/error-handling), we explore how destructuring values aids in error handling when using `try...with`.
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -209,15 +214,103 @@ val forename : string = "Robin"
+
+Pattern matching also works with nested user-defined types.
+
+In the example below, we deconstruct nested tuples:
```
```suggestion
Pattern matching also works with nested user-defined types. In the example below, we deconstruct nested tuples:
```

Avoid single sentence paragraphs

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,103 @@ val forename : string = "Robin"
+```ocaml
+# let () =
+  Printf.printf "Name: %s\nStreet: %s\nCity: %s\nZip: %d\nEmail: %s\nPhone: %d\n"
+    name street city zip email phone;;
````
```suggestion
# Printf.printf "Name: %s\nStreet: %s\nCity: %s\nZip: %d\nEmail: %s\nPhone: %d\n"
    name street city zip email phone;;
```

`let () =` isn't needed in interactive top level

bbatsov commented:
To be more specific - I'm suggesting to add something like:

```
name: Markdown Lint

on:
  pull_request:
    paths:
      - '**/*.md'

jobs:
  lint-markdown:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3

      - name: Get changed files
        id: changed-files
        uses: tj-actions/changed-files@v41
        with:
          files: '**/*.md'

      - name: Lint Markdown files
        uses: DavidAnson/markdownlint-cli2-action@v13
        with:
          globs: ${{ steps.changed-files.outputs.all_changed_files }}
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -209,15 +214,103 @@ val forename : string = "Robin"
+# type address = {
+  street: string;
+  city: string;
+  zip: int
```
```suggestion
  street : string;
  city : string;
  zip : int
```

cuihtlauac reviewed (changes requested):
Thanks @jakubsvec001. Nested patterns and discarding names were missing in the tutorial.

Do you think it is possible to remove the `address` type throughout your example?  We try to minimize the text lengths and focus examples on a single idea, a single thing we want to highlight. That makes the reading and learning experience easier.

Let me know if I've overlooked something you wanted to explain.

cuihtlauac commented:
@bbatsov Can you move the  `markdownlint-cli2-action` discussion into a separate PR, draft if needed? :heart: Let's try to have one topic per issue/PR, otherwise it make search and long term maintenance difficult.

cuihtlauac commented:
@jakubsvec001: Can you move your workflow comments in ad-hoc issue. This is important topic, your suggestions are valuable. We'd like them to appear on their own in the tracker 🙏

jakubsvec001 committed:
Update data/tutorials/language/0it_00_values_functions.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

jakubsvec001 committed:
Update data/tutorials/language/0it_00_values_functions.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

jakubsvec001 committed:
Update data/tutorials/language/0it_00_values_functions.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

jakubsvec001 committed:
Update data/tutorials/language/0it_00_values_functions.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

jakubsvec001 committed:
Update data/tutorials/language/0it_00_values_functions.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

jakubsvec001 commented:
> Do you think it is possible to remove the `address` type throughout your example? We try to minimize the text lengths and focus examples on a single idea, a single thing we want to highlight. That makes the reading and learning experience easier.
> 
> Let me know if I've overlooked something you wanted to explain.

Happy to do this. I will update this later this morning.

jakubsvec001 committed:
Replace overly complicated nested deconstruction example with simpler version

jakubsvec001 commented:
@cuihtlauac I removed the "address" field from the record of the nested deconstruction example

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -114,6 +114,11 @@ Arbitrary combinations of chaining or nesting are allowed.
+
+[Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) is a programming language construct that generalizes case analysis, it makes subexpression inspection possible and applies to values of any type. 
+
+In the following sections, we explore matching inside `let` bindings. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in the construction of conditional expressions using `match...with` and compare it to `if...then...else`. In the chapter on [Error Handling](/docs/error-handling), we explore how destructuring values aids in error ...
```
```suggestion
In the following sections, we explore matching inside `let` bindings, which is a special case. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in general cases using `match...with`. The chapter on [Error Handling](/docs/error-handling) explores how destructuring values aids in error handling when using `try...with`.
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -114,6 +114,11 @@ Arbitrary combinations of chaining or nesting are allowed.
 
 In both examples, `d` and `e` are local definitions.
 
+## Introduction to Pattern Matching
```
```suggestion
## Forms of Pattern Matching
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,124 @@ val forename : string = "Robin"
 ```
 
-The `List.split` function returns a pair of lists. We're only interested in the second list, we give it the name `y` and discard the first list by using `_`.
+Tuples behave differently from records because they require positional consistency. To discard the `email` value from the tuple of the `contact` field, we need to use the catch-all pattern (`_`):
````
```suggestion
Tuples behave differently from records; contained data is anonymous, and its position is used to access it. To discard the `email` value from the tuple of the `contact` field, we need to use the catch-all pattern (`_`):
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -209,15 +214,124 @@ val forename : string = "Robin"
+
+In the following example, we deconstruct a tuple nested within a record.
+
+Records require explicit type definitions wherein field names are annotated with field types. We first define a `person` record type, then create an instance of that type named `jane`:
```
```suggestion
Records require explicit type definitions wherein field names are annotated with types. We first define a `person` record type, then create an instance of that type named `jane`:
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,124 @@ val forename : string = "Robin"
+   contact = ("jane@example.com", 1234567890)}
+```
+
+The following examples demonstrate two methods by which we can extract Jane's `email` and `phone` number data in the nested `contact` tuple. We will do so first by using nested deconstruction, then demonstrate a brute force approach by first extracting `content` followed by accessing the `email` and `phone` by deconstructing `contact`. <!-- Note: This sequence is pedagologically backwards perhaps ...
````
```suggestion
The following examples demonstrate two methods by which we can extract Jane's `email` and `phone` number data in the nested `contact` tuple. We will do so first by using nested deconstruction, then demonstrate another approach by first extracting `content` followed by accessing the `email` and `phone` by deconstructing `contact`. 
```
>  Note: This sequence is pedagologically backwards perhaps, but the goal is to show that `contact` does not become a bound variable when using nested deconstruction. This is difficult to demonstrate in a single utop session if `contact` is bound in the brute force approach first, as it will not show the desired error message

Nice suggestion, thanks

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
````diff
@@ -209,15 +214,124 @@ val forename : string = "Robin"
+Error: Unbound value contact
+```
+
+This is because "contact" is not a bound variable, rather it is an intermediate pattern for deconstructive pattern-matching.
````
```suggestion
This is because "contact" is not a top-level definition; rather, it is a local definition from deconstructive pattern-matching.
```

cuihtlauac commented on data/tutorials/language/0it_00_values_functions.md:
```diff
@@ -209,15 +214,124 @@ val forename : string = "Robin"
+
+This is because "contact" is not a bound variable, rather it is an intermediate pattern for deconstructive pattern-matching.
+
+Next, we demonstrate the brute force approach for accessing `email` and `phone`. This brings `contact` into our top-level scope and requires two steps:
```
```suggestion
Next, we demonstrate the two-phase approach for accessing `email` and `phone`. This brings `contact` into our top-level scope and requires two steps:
```

cuihtlauac reviewed (changes requested):
Thanks, @jakubsvec001, this is great. Sorry for the delay reviewing. I'm suggesting some additional changes.

cuihtlauac commented:
> To be more specific - I'm suggesting to add something like:
> 
> ```
> name: Markdown Lint
> 
> on:
>   pull_request:
>     paths:
>       - '**/*.md'
> 
> jobs:
>   lint-markdown:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: actions/checkout@v3
> 
>       - name: Get changed files
>         id: changed-files
>         uses: tj-actions/changed-files@v41
>         with:
>           files: '**/*.md'
> 
>       - name: Lint Markdown files
>         uses: DavidAnson/markdownlint-cli2-action@v13
>         with:
>           globs: ${{ steps.changed-files.outputs.all_changed_files }}
> ```

@bbatsov could you turn that suggestion into a PR please, that sounds great

bbatsov commented:
Will do!

bbatsov commented:
@cuihtlauac See #3012 and #3013.

jakubsvec001 committed:
Apply suggestions from code review

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

jakubsvec001 committed:
Update data/tutorials/language/0it_00_values_functions.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>
